ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -130,19 +130,32 @@ public AmbryRequests(StoreManager storeManager, RequestResponseChannel requestRe
     this.replicationEngine = replicationEngine;
     this.notification = operationNotification;
     this.storeKeyFactory = storeKeyFactory;
-    this.storeKeyConverterFactory = storeKeyConverterFactory;
     this.connectionPool = connectionPool;
     this.metricRegistry = registry;
     this.serverConfig = serverConfig;
-    transformer = null;
-    if (serverConfig != null) {
+
+    keyConverter = ThreadLocal.withInitial(()-> {
       try {
-        StoreKeyConverter keyConverter = storeKeyConverterFactory.getStoreKeyConverter();
-        transformer = Utils.getObj(serverConfig.serverMessageTransformer, storeKeyFactory, keyConverter);
+        return storeKeyConverterFactory.getStoreKeyConverter();
       } catch (Exception e) {
-        logger.error("Failed to create transformer", e);
+        logger.error("Failed to create keyConverter!!!", e);
       }
-    }
+      return null;
+    });
+
+    /* All the request handlers share one single AmbryRequests object.
+     * But the StoreKeyConverter of the Transformer has a cache which shouldn't be shared among handlers.
+     * So use ThreadLocal transformer. */
+    this.transformer = ThreadLocal.withInitial(() -> {
+      if (serverConfig != null && keyConverter != null && keyConverter.get() != null) {
```

Comment:
we don't have to check if the keyConverter is null, we know it's not going to be null.